packages/next/src/app-dir/links/nextCache.ts

```diff
@@ -1,8 +1,12 @@
 // import "server-only";
 
 import { TRPCClientError, TRPCLink } from '@trpc/client';
-import { AnyRouter, callProcedure, inferRouterContext } from '@trpc/server';
 import { observable } from '@trpc/server/observable';
+import {
+  AnyRouter,
+  callProcedure,
```

Comment:
We should have a public way of doing this as it's nice for RSC

Reply:
Don't you use `router.createCaller()` for that?

Reply:
That said, we should deprecate `router.createCaller` and create a `createCaller(x)` instead

Reply:
createCaller doesnt play very nice as a client if you have async context.

createTRPCClient with custom link using callProcedure is the best I've found

Reply:
Could you add an example of this within the tRPC repo somewhere so I can understand it better?

Reply:
This shows pretty well why it's not the best ergonomics when using createCaller.

https://github.com/t3-oss/create-t3-turbo/blob/eae39060ee1ea1bad06842c3a3ce7a80463405c5/apps/nextjs/src/trpc/server.ts#L63-L72

The two accomplishes the exact same thing but the client doesn't require nested awaits.

We could probably make something that's even better though, like allowing createCaller to accept a function, and then that'd be called at the same time as the actual procedure. Since you can wrap it in cache, it'd still be de-duped.

```ts
const trpc = appRouter.createCaller(createContext);

const post = await trpc.post.all();
```

Reply:
although you'd still miss things like having the loggerLink etc